# Megaline Plan Reccomendation.

## Open Data and Study

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/users_behavior.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [2]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Here i displayed the information and the first five rows of the dataset to understand which parameters will be used to make the prediction and which value we are trying to predict. I found that the first four columns are the data we will used to preedict the fifth which is the plan that we will reccomend to the consumer.

## Split Dataset and Define Parameters

In [3]:
from sklearn.model_selection import train_test_split
df_train, df_valid = train_test_split(df, test_size=0.4, random_state=12345)
df_valid, df_test = train_test_split(df_valid, test_size=0.5, random_state=12345)

features = df_train[['calls', 'minutes', 'messages', 'mb_used']]
target = df_train['is_ultra']

valid_features = df_valid[['calls', 'minutes', 'messages', 'mb_used']]
valid_target = df_valid['is_ultra']

test_features = df_test[['calls', 'minutes', 'messages', 'mb_used']]
test_target = df_test['is_ultra']

In this cell I divided the data into a training set, a test set, and a validation, with a 3:1:1 ratio set by using the train_test_split method. i then crated the features and target variables that i will use to train, validate and test the model using the parameters i identified in the previous task.

## Train Models and Evaluate Quality

In [4]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

for depth in range(1, 6):
        dt_model = dt_model = DecisionTreeClassifier(max_depth=depth)

        dt_model.fit(features, target)

        dt_predictions_valid = dt_model.predict(valid_features)

        print("max_depth =", depth, ": ", end='')
        print(accuracy_score(valid_target, dt_predictions_valid))


max_depth = 1 : 0.7542768273716952
max_depth = 2 : 0.7822706065318819
max_depth = 3 : 0.7853810264385692
max_depth = 4 : 0.7791601866251944
max_depth = 5 : 0.7791601866251944


In [5]:
from sklearn.ensemble import RandomForestClassifier

for estimators in range(1, 6):
        rf_model = rf_model = RandomForestClassifier(random_state=12345, n_estimators=estimators)

        rf_model.fit(features, target)

        rf_predictions_valid = rf_model.predict(valid_features)

        print("n_estimators =", estimators, ": ", end='')
        print(accuracy_score(valid_target, rf_predictions_valid))

n_estimators = 1 : 0.7107309486780715
n_estimators = 2 : 0.7636080870917574
n_estimators = 3 : 0.7387247278382582
n_estimators = 4 : 0.7713841368584758
n_estimators = 5 : 0.749611197511664


In [6]:
from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression(random_state=12345)
lr_model.fit(features, target)
lr_valid_predictions = lr_model.predict(valid_features)
print(accuracy_score(valid_target, lr_valid_predictions))


0.7589424572317263


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


I tested three different models using the validation set and by changing hyperparameters using a for loop to find the best model. The best model seems to be the decision tree model because it had the best accuracy score when changing the hyperparameters. The accuracy score is the highest and the decision tree model is one of the faster models this will allow us the deal out plan reccomendations as quickly as possible. Therefore, I will move forward with the decision tree model and test its accuracy using the test set.

## Test Chosen Models

In [7]:
rf_test_predictions = rf_model.predict(test_features)
print(accuracy_score(test_target, rf_test_predictions))

0.7869362363919129


The decision tree model preformed well. It predicted plans with over 78% accuracy. This means that by using this model 78% of the customers that use it will get the proper reccomendation for the plan that suits their needs. By being a decision tree model the reccomendation will be fast and the customer can fill out parameters by questionaire.